pydra/engine/core.py

```diff
@@ -700,8 +700,9 @@ def add(self, task):
             raise Exception("Submitter should already be set.")
         # at this point Workflow is stateless so this should be fine
         nwf = await submitter.submit(self, return_task=True)
-        # TODO: check if this might be done differently
-        self.__dict__.update(nwf.__dict__)
+        # dj TODO: don't understand why I need it
+        if not self.state:
+            self.__dict__.update(nwf.__dict__)
```

Comment:
can you check what updates are propagated? that may give a clue.

Reply:
mostly graph

Reply:
why is `graph` relevant to be updated?

Reply:
I mean - I was checking what is different between `self.__dict__` and `nwf.__dict__`, and it was about the object from the graph, but could be more specific:

This are things that differ:
```
for key in ["inputs", "graph", 'name2obj', '_connections', '_last_added']: key, self.__dict__[key], nwf.__dict__[key]
('inputs', Inputs(x=2, _graph=[add2]), Inputs(x=2, _graph=[add2]))
('graph', <networkx.classes.digraph.DiGraph object at 0x7fd3ed08e6d8>, <networkx.classes.digraph.DiGraph object at 0x7fd3ed0c38d0>)
('name2obj', {'add2': add2}, {'add2': add2})
('_connections', [('out', LF('add2', 'out'))], [('out', LF('add2', 'out'))])
('_last_added', add2, add2)
```

And I've just check the inputs for `add2`, and one has set input `x=2`, the other doesn't:
```
(Pdb) nwf.__dict__['name2obj']["add2"].inputs.x
2
(Pdb) self.__dict__['name2obj']["add2"].inputs.x
LF('wf_1', 'x')
```

Reply:
I've just realized that in this method there is never `runnable` without state...